src/distdl/nn/mixins/halo_mixin_new.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+
+from distdl.utilities.slicing import compute_start_index
+from distdl.utilities.slicing import compute_subshape
+
+
+class HaloMixin:
+
+    def _compute_exchange_info(self,
+                               halo_shape,
+                               neighbor_halo_shapes):
+
+        recv_buffer_shape = halo_shape.copy()
+        send_buffer_shape = np.zeros_like(halo_shape, dtype=int)
+
+        dims = len(halo_shape)
+        for dim in range(dims):
+            left_halo_shape, right_halo_shape = neighbor_halo_shapes[dim]
+
+            if left_halo_shape is not None:
+                send_buffer_shape[dim, 0] = left_halo_shape[dim, 1]
+
+            if right_halo_shape is not None:
+                send_buffer_shape[dim, 1] = right_halo_shape[dim, 0]
+
+        return recv_buffer_shape, send_buffer_shape
+
+    def _left_pad_to_length(self, array, length, value):
+        lpad = length - len(array)
+        return np.pad(array,
+                      pad_width=(lpad, 0),
+                      mode='constant',
+                      constant_values=value)
+
+    def _compute_local_start_stop_indices(self, subtensor_shapes_unbalanced, x_local_shape):
+        # this could use a refactor at some point
+        x_local_start_index = np.zeros_like(x_local_shape, dtype=np.int)
+        dims = len(x_local_shape)
+        for dim in range(2, dims):
```

Comment:
We should be able to write this in such a way that it does not depend on ignoring the channel and batch dimensions.

Reply:
@rhewett Yes, that loop can go from [0, dims) and still be correct. Will fix.

@thomasjgrady, if we try to use `compute_subtensor_[start|stop]_indices`, then we would have to either...
- perform a communication in HaloMixin
- pass in the left/right neighbor start/stop indices per dimension (or just all of them)

In either case, an extra communication is required.